src/main/scala/net/snowflake/spark/snowflake/TelemetryReporter.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2015-2020 Snowflake Computing
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.snowflake.spark.snowflake
+
```

Comment:
To reviewers:
1. TelemetryReporter is a trait, there are 3 concrete implementations: NoopTelemetryReporter, DriverTelemetryReporter and ExecutorTelemetryReporter.
  a) NoopTelemetryReporter does nothing, it is set by default for both driver and executor.
  b) DriverTelemetryReporter wrap all the details to send telemetry on driver side. If the feature is enabled, entry-functions in DefaultSource e.g. createRelation() will set up DriverTelemetryReporter instance. So all LoggerWrapper instance in driver side can use it to send log message.
  c) ExecutorTelemetryReporter wrap all the details to send telemetry on executor side. If the feature is enabled, driver will pass JDBC session token to executors. In the begin of each task, it sets up DriverTelemetryReporter instance. So all LoggerWrapper instances in executor can use it to send log message.

2. How TelemetryReporter.getTelemetryReporter() will get an appropriate instance?
a) "object TelemetryReporter" will have a singleton instance for driver TelemetryReporter
b) "object TelemetryReporter" will have a ThreadLocal instance for executor TelemetryReporter
c) "object TelemetryReporter" will have a ThreadLocal instance of Boolean to indicate whether the thread is executor.
d) If the thread is an executor, getTelemetryReporter() returns the TelemetryReport for executor, otherwise, it returns the reference of driver TelemetryReporter